core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -158,6 +159,8 @@
 
 public class ExpressionInterpreter
 {
+    private static final Map<InListExpression, Set<?>> inListCache = new WeakHashMap<>();
```

Comment:
FYI: we generally do not use static state like this, so this requires a discussion

Reply:
Wondering what's wrong with using a static state? I tried making ExpressionInterpreter long-lived but this requires a lot of changes since it is being used in many different places.

One option is if we make ExpressionInterpreter singleton and pass all the params (expression, plannerContext) directly in the optimize or evaluate method.

Reply:
> Generally static mutable state is much harder to test.

Yes, testing would be an issue

Reply:
Cache shouldn't be static. It should be part of planning context for a query. Some options:
1. Currently, for iterative rules we have `io.trino.sql.planner.iterative.Rule.Context`. Perhaps we should have something similar for `PlanOptimizer`, e.g: `PlanOptimizer#Context`. Then you could provide an interface for optimizing expression there or just keep per-query `ExpressionInterpreter`
2. We already `PlannerContext` object, but it's global. We could make local to the query. However, I think will be more involved as there are other global objects that depend on it.

I'm leaning towards 1)

cc @raunaqmorarka

Reply:
Implemented as part of second commit